enhancements/catalog-source-config.md

```diff
@@ -0,0 +1,340 @@
+
+```yaml
+spec:
+  type: object
+  required:
+    - sourceConfig
+  properties:
+    # Deprecate 
+    # image: ... 
+    # address: ... 
+    # configMap: ...
+    # sourceType: grpc | configMap | internal
+    
+    # Keep as is:
+    description: ...
+    displayName: ...
+    icon: ...
+    priority: ...
+    publisher: ...
+    secrets: ...
+
+    # Add:
+    # Note: 'internal' type is being deprecated so it gets no config structure
+    grpcPodSource:
+      podSpec:  # a subset (or the whole of) corev1.PodSpec
+        image: ... # required
+        # all other attributes are optional
+        # and override the default catalog source pod spec
+        ...
+      ...
+    grpcServiceSource:
+      address: ...
+      ...
+    configMapSource:
+      configMap: ...
+      ... 
+  oneOf:
+    - required: ["grpcPodSource"]
+    - required: ["grpcServiceSource"]
+    - required: ["configMapSource"]
+                  
+``` 
+
+This option introduces breaking changes. This means that while we are on the depracation path, the users will be able to configure their `CatalogSource`s in the current way, _OR_, use one of `grpcPodSource`, `grpcServiceSource`, or `configMapSource` (with precendence given to the new configuration). 
+
+**Pros**
+  1. Each CatalogSource type has its own configuration structure
+  2. We don't polute the top level with endless optional attributes that behave differently depending on the values of other attributes
+  3. Better UX
+  4. New catalog source types can be added easily
+
+
+**Cons**
+  1. Introduces breaking changes
+  2. There is not first-class support for union type validation in kubebuilder/operator-sdk. We'll need another mechanism to add and test these validations (e.g. patch that gets executed after `make controller-gen` in the `api` repo)
+
+#### CatalogSourceClass
+
+In this option, we transion the `CatalogSource` to a more plugable architecture by introducing a `catalogSourceClass` and a schemaless `parameters` block where arbitrary parameters can be passed in. The controller would then be in charge of reconciling catalog sources of classes it recognizes.
+
+In the future, we'd have something like this:
+
+```yaml
+apiVersion: operators.coreos.com/v1alpha2
+kind: CatalogSource
+metadata:
+  name: operatorhubio-catalog
+  namespace: olm
+spec:
+  catalogSourceClass: grpc-pod
+  displayName: Community Operators
+  publisher: OperatorHub.io
+  parameters:
+    image: quay.io/operatorhubio/catalog:latest
+    # additional parameters are options and override the underlying podspec
+```
+
+**Pros**
+  1. Extensible architecture: users can provide their own `CatalogSourceClass` types and controllers
+  2. Better UX than the status quo: top level attributes are shared and `parameters` should only contain information relevant for that class of catalog source
+
+**Cons**
+  1. `parameters` is schemaless. So, we'd also need an adminission webhook
+  2. Requires deprecation
+
+#### ImageConfig
```

Comment:
I am surprised by the choice of the name: ImageConfig. I may be missing something as I don't see the relation with the container image. I would have expected more something like affinity or scheduling similar to the pod structure:
https://kubernetes.io/docs/reference/kubernetes-api/workload-resources/pod-v1/#scheduling

Reply:
Yeah, maybe `ImageConfig` is a bad name. I'll address this in the next iteration. Maybe `PodConfig`, or `PodSpecOverrides` or something like that. We may not want to give the user the option to override every aspect of the pod spec right out of the gate, but we should keep the name general enough to encompasse more behaviours if need be.

Reply:
The problem with the `CatalogSource` API in this context is that a bunch of unrelated fields are munged together, such that some of them are inert depending on the values of `sourceType` and `image`


| `sourceType` | `image`  | `address` | `updateStrategy` | `configMap` |
| ----------------- | ----------- | ------------- | ---------------------- | ---------------- |
| `grpc`  | non-empty  | inert | active | inert |
| `grpc` | empty | active | inert | inert |
| `configmap` | inert | inert | inert | active |
| `internal` | ? | ? | ? | ? |

If we add another root field that specifies how to configure the pod when that catalog source references an image, it will be inert in all but the first case.

IMO `ImageConfig` is a tip of the hat to this reality to make it more obvious to users when its active. Not an ideal name, but I can see the logic in it.

Reply:
@joelanford seems like you're describing a union (i.e. "sum") type

Reply:
That would be a solution to the problem I'm describing, but that would mean a significant change to the `CatalogSource` API, likely involving a new version and a conversion webhook, which starts to look somewhat scary just to be able to set a few more pod defaults.

Reply:
Right! IMO this is the right way to do it. The question to the community is whether it is worth the effort. Personally, I think it gives a better user experience and level of polish. It's totally valid to ask if it's worth going through the hassle of a version change right *now*. Since this change doesn't add much to many users, maybe we can continue with the status quo and "save" doing it right for a bigger change. This is what I need feedback on. If we don't have strong feelings, then my suggestion is the following:
1. Do it the quick and dirty way (just add another optional structure)
2. Tee up the refactoring of the API structure for a particular release. I'll finish this EP and detail the transition.